src/gencost/waterfall.py

```diff
@@ -2256,3 +2267,480 @@ def valid_generation(df_):
+                    x["ba_plus_essentials"].isin(hist_data["ba_plus_essentials"]),
+                    1,
+                    0,
+                ),
+                ba_plus_age_present=lambda x: np.where(
+                    x["ba_plus_age"].isin(hist_data["ba_plus_age"]),
+                    1,
+                    0,
+                ),
+            )  # make an exception of when there is no ba code or age
+            .assign(
+                ba_plus_age_present=lambda x: np.where(
+                    (x["final_ba_code"].isnull())
+                    | (x["final_ba_code"].isna())
+                    | (x["age_in_current_year"].isna())
+                    | (x["age_in_current_year"].isnull()),
+                    0,
+                    x["ba_plus_age_present"],
+                ),
+                ba_plus_essentials_present=lambda x: np.where(
+                    (x["final_ba_code"].isnull()) | (x["final_ba_code"].isna()),
+                    0,
+                    x["ba_plus_essentials_present"],
+                ),
+                fill_in_score=lambda x: x["essentials_present"]
+                + x["ba_plus_essentials_present"]
+                + x["ba_plus_age_present"],
+            )
+        )
+
+        # some sort of for loop that sets prioritizes merging based on fill in score
+
+        three = (
+            missing_data.query("fill_in_score ==3")
+            .merge(
+                hist_data.drop(
+                    columns=[
+                        "plant_id_eia",
+                        "generator_id",
+                        "year",
+                        "prime_mover",
+                        "fuel_group",
+                        "final_ba_code",
+                        "age_in_current_year",
+                        "age_range",
+                        "essentials",
+                        "ba_plus_essentials",
+                    ]
+                ),
+                on=["ba_plus_age"],
+                how="left",
+                indicator=True,
+            )
+            .query('_merge == "both"')
+            .drop_duplicates(subset=["plant_id_eia", "generator_id", "year"])
+        )
+
+        two = (
+            missing_data.query("fill_in_score ==2")
+            .merge(
+                hist_data.drop(
+                    columns=[
+                        "plant_id_eia",
+                        "generator_id",
+                        "year",
+                        "prime_mover",
+                        "fuel_group",
+                        "final_ba_code",
+                        "age_in_current_year",
+                        "age_range",
+                        "ba_plus_age",
+                        "essentials",
+                    ]
+                ),
+                on=["ba_plus_essentials"],
+                how="left",
+                indicator=True,
+            )
+            .query('_merge == "both"')
```

Comment:
You can do this merge and the downselection to 'both' by setting `how` to 'inner'